File: ios/RoutineChart/RoutineChart/Core/Domain/Repositories/AuthRepository.swift
//
//  AuthRepository.swift
//  RoutineChart
//
//  Created for Phase 2.1: Firebase Auth
//

import Foundation
import Combine

/// Protocol for authentication operations
protocol AuthRepository {
    /// Current authenticated user (nil if not authenticated)
    var currentUser: AuthUser? { get }
    
    /// Publisher that emits when auth state changes
    var authStatePublisher: AnyPublisher<AuthUser?, Never> { get }
    
    /// Sign in with email and password (parent flow)
    func signInWithEmail(email: String, password: String) async throws -> AuthUser
    
    /// Sign up with email and password (parent flow)
    func signUpWithEmail(email: String, password: String) async throws -> AuthUser
    
    /// Sign in anonymously (child flow)
    func signInAnonymously() async throws -> AuthUser
    
    /// Link anonymous account to email (upgrade child to parent)
    func linkAnonymousToEmail(email: String, password: String) async throws -> AuthUser
    
    /// Sign out
    func signOut() throws
    
    /// Send password reset email
    func sendPasswordReset(email: String) async throws
}

File: android/app/src/main/java/com/HammersTech/RoutineChart/core/domain/repositories/AuthRepository.kt
package com.HammersTech.RoutineChart.core.domain.repositories

import com.HammersTech.RoutineChart.core.domain.models.AuthUser
import kotlinx.coroutines.flow.Flow

/**
 * Repository interface for authentication operations
 * Phase 2.1: Firebase Auth
 */
interface AuthRepository {
    /**
     * Flow of the current authenticated user (null if not authenticated)
     */
    val authStateFlow: Flow<AuthUser?>

    /**
     * Current authenticated user (null if not authenticated)
     */
    val currentUser: AuthUser?

    /**
     * Sign in with email and password (parent flow)
     */
    suspend fun signInWithEmail(
        email: String,
        password: String,
    ): Result<AuthUser>

    /**
     * Sign up with email and password (parent flow)
     */
    suspend fun signUpWithEmail(
        email: String,
        password: String,
    ): Result<AuthUser>

    /**
     * Sign in anonymously (child flow)
     */
    suspend fun signInAnonymously(): Result<AuthUser>

    /**
     * Link anonymous account to email (upgrade child to parent)
     */
    suspend fun linkAnonymousToEmail(
        email: String,
        password: String,
    ): Result<AuthUser>

    /**
     * Sign out
     */
    suspend fun signOut(): Result<Unit>

    /**
     * Send password reset email
     */
    suspend fun sendPasswordReset(email: String): Result<Unit>
}
